backend/scripts/test_citation_builder.py

```diff
@@ -0,0 +1,281 @@
+
+# pylint: enable=E402
+# flake8: noqa: E402
+import random
+
+
+def create_mock_tokens(text: str) -> list[str]:
+    tokens = []
+    i = 0
+    while i < len(text):
+        # Randomly choose to split after 2, 3, or 4 characters
+        split_length = random.choice([2, 3,4, 5, 6])
+        end = min(i + split_length, len(text))
+        tokens.append(text[i:end])
+        i = end
+    return tokens
+
+
+
+def create_mock_doc_id_to_rank_map(num_docs: int) -> dict[str, int]:
+    return {f"doc_{i}": i + 1 for i in range(num_docs)}
+
+
+mock_docs = [
+    LlmDoc(
+        document_id=f"doc_{int(id/2)}",
+        content="Document is a doc",
+        blurb=f"Document #{id}",
+        semantic_identifier=f"Doc {id}",
+        source_type=DocumentSource.WEB,
+        metadata={},
+        updated_at=datetime.now(),
+        link=f"https://{int(id/2)}.com" if int(id / 2) % 2 == 0 else None,
+        source_links={0: "https://mintlify.com/docs/settings/broken-links"},
+    )
+    for id in range(10)
+]
+
+
+mock_doc_mapping = {
+    "doc_0": 0,
+    "doc_1": 1,
+    "doc_2": 2,
+    "doc_3": 3,
+    "doc_4": 4,
+    "doc_5": 5,
+}
+
+
+@pytest.fixture
+def mock_data():
+    return mock_docs, mock_doc_mapping
+
+
+def print_output(output: List[Union[DanswerAnswerPiece, CitationInfo]]):
+    for item in output:
+        if isinstance(item, DanswerAnswerPiece):
+            print(f"DanswerAnswerPiece: {item.answer_piece}")
+        elif isinstance(item, CitationInfo):
+            print(
+                f"CitationInfo: citation_num={item.citation_num}, document_id={item.document_id}"
+            )
+
+
+def are_outputs_equivalent(output1, output2):
+    if len(output1) != len(output2):
+        return False
+    for item1, item2 in zip(output1, output2):
+        if type(item1) != type(item2):
+            return False
+        if isinstance(item1, DanswerAnswerPiece):
+            if item1.answer_piece.strip() != item2.answer_piece.strip():
+                return False
+        elif isinstance(item1, CitationInfo):
+            if (
+                item1.citation_num != item2.citation_num
+                or item1.document_id != item2.document_id
+            ):
+                return False
+    return True
+
+
+def process_text(input_text, mock_data):
+    mock_docs, mock_doc_id_to_rank_map = mock_data
+
+    mock_tokens = create_mock_tokens(input_text)
+    result = list(
+        extract_citations_from_stream(
+            tokens=iter(mock_tokens),
+            context_docs=mock_docs,
+            doc_id_to_rank_map=mock_doc_id_to_rank_map,
+            stop_stream=None,
+        )
+    )
+    final_answer_text = ""
+    citations = []
+    for piece in result:
+        if isinstance(piece, DanswerAnswerPiece):
+            final_answer_text += piece.answer_piece
+        elif isinstance(piece, CitationInfo):
+            citations.append(piece)
+    return final_answer_text, citations
+
+
+import argparse
```

Comment:
Imports can be top of file but we don't want to have argparse for pytests generally

For typing List, let's just use the built in list

Reply:
All great feedback! Sorry for leaving the unused logs / non-best practices in- was planning on cleaning this up before submitting as a PR (still using the non-deterministic tokenizing to thoroughly test some tricky edge cases). 

Completely agree with the comments and will make the modifications!